pkg/jx/cmd/controller_build.go

```diff
@@ -321,6 +328,116 @@ func (o *ControllerBuildOptions) createPromoteStepActivityKey(buildName string,
 	}
 }
 
+// completeBuildSourceInfo sets the PR author and PR title from GitHub in the given PA
+// If the PA is a branch build it then sets the commit author and last commit message
+func (o *ControllerBuildOptions) completeBuildSourceInfo(activity *v1.PipelineActivity) {
+
+	log.Infof("[BuildInfo] Completing build info for PipelineActivity=%s\n", activity.Name)
+
+	gitInfo, err := gits.ParseGitURL(activity.Spec.GitURL)
+	if err != nil {
+		log.Warnf("Cannot parse git repo URL [%s]. Error: %s\n", activity.Spec.GitURL, err)
+		return
+	}
+	if !gitInfo.IsGitHub() {
+		// this is GH only for now
+		return
+	}
+	if activity.Spec.Author != "" {
+		// info already set, save some GH requests
+		return
+	}
+
+	secrets, err := o.LoadPipelineSecrets(kube.ValueKindGit, "github")
+	if err != nil {
+		log.Warnf("Cannot load pipeline secrets. Error: %s\n", err)
+	}
+
+	// get a github API client
+	provider, err := o.getGithubProvider(secrets, gitInfo)
+	if err != nil {
+		log.Warnf("Cannot create GitHub provider. Error: %s", err)
+	}
+
+	// extract (org, repo, commit) or (org, repo, #PR) from key
+	if strings.HasPrefix(strings.ToUpper(activity.Spec.GitBranch), "PR-") {
+		// this is a PR build
+		n := strings.Replace(strings.ToUpper(activity.Spec.GitBranch), "PR-", "", -1)
+		prNumber, err := strconv.Atoi(n)
+		if err != nil {
+			log.Warnf("Cannot extract PR number from [%s]. Error: %s\n", activity.Spec.GitBranch, err)
+			return
+		}
+		pr, e := provider.GetPullRequest(gitInfo.Organisation, gitInfo, prNumber)
+		if e != nil {
+			log.Warnf("Cannot read PR from Github. Error: %s\n", e)
+			return
+		}
+		if pr.Author != nil {
+			activity.Spec.Author = pr.Author.Login
+		}
+		activity.Spec.PullTitle = pr.Title
+		log.Infof("[BuildInfo] PipelineActivity set with author=%s and PR title=%s\n", activity.Spec.Author, activity.Spec.PullTitle)
+	} else {
+		// this is a branch build
+		gitCommits, e := provider.ListCommits(gitInfo.Organisation, gitInfo.Name, &gits.ListCommitsArguments{
+			SHA:     activity.Spec.GitBranch,
+			Page:    1,
+			PerPage: 1,
+		})
+		if e != nil {
+			log.Warnf("Cannot read last branch commit from Github. Error: %s\n", e)
+			return
+		}
+		if len(gitCommits) > 0 {
+			if gitCommits[0] != nil && gitCommits[0].Author != nil {
+				activity.Spec.Author = gitCommits[0].Author.Login
+				activity.Spec.LastCommitMessage = gitCommits[0].Message
+			}
+		}
+		log.Infof("[BuildInfo] PipelineActicity set with author=%s and last message=%s\n", activity.Spec.Author, activity.Spec.LastCommitMessage)
+	}
+
+}
+
+func (o *ControllerBuildOptions) getGithubProvider(secrets *corev1.SecretList, gitInfo *gits.GitRepository) (gits.GitProvider, error) {
+	// this internal provider is only used during tests
+	if o.gitHubProvider != nil {
+		return o.gitHubProvider, nil
+	}
+
+	// production code always goes this way
+	if secrets != nil {
```

Comment:
This is specific only to GitHub. 

I would consider using instead of this block of code: `CommonOptions.getPipelineGitAuth` + `gits.CreateProvider` which abstracts away the provider type.

Reply:
I intentionally implemented this for `github.com` only as it is the only fully supported git provider. Also, I don't think it makes sense make the secrets part generic and not the rest (info retrieval is using the GH API client). Making this really generic would require a lot more work.

Reply:
It makes sense, but please use the helper methods instead of implementing this on your own. It is less maintains and the risk to break things is lower. Thanks

Reply:
I see, you want to abstract the secrets retrieval so any change on secrets storage is not impacting this. Got it. Fixing.